### PR TITLE
Make generic

### PR DESCRIPTION
**Pull request by troyxli**
Remove hard-coded types

### CODE REVIEW

File: RuleEngine/Application.cs [hunk 1]
```diff
@@ -7,13 +7,28 @@
...
 {
-	public class Application
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public class Application
 	{
-		private IParser _parser;
+        private static readonly Dictionary<string, Type> _knownTypes =
+            new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
```
Comment (troyxli) [inline RuleEngine/Application.cs:16]: Reflection does not work for primitive types. So I have to map them manually. See the static constructor.

File: RuleEngine/Application.cs [hunk 2]
```diff
@@ -38,27 +53,49 @@ public void Run(string[] args)
...
-				{
-					doubleResult = _parser.Parse<double>(rule.Expression);
-					_log.LogInformation($"Result of expression {rule.Expression} is {doubleResult}");
+                var returnType = GetShortcutType(rule.ReturnType) ?? Type.GetType(rule.ReturnType, false, true);
+                var methods = typeof(IParser).GetMethods().Where(m => m.ContainsGenericParameters).ToList();
+
+                if (string.IsNullOrWhiteSpace(rule.InputType))
+                {
+                    var method = methods.First(
+                        m => m.ReturnType.IsGenericParameter && m.GetGenericArguments().Length == 1 &&
```
Comment (troyxli) [inline RuleEngine/Application.cs:65]: Not sure this is the best way to find the generic method V Parse<V>(string expression)

File: RuleEngine/Application.cs [hunk 3]
```diff
@@ -38,27 +53,49 @@ public void Run(string[] args)
...
+                    _log.LogInformation($"Result of expression {rule.Expression} is {finalResult}");
 				}
 				else
-				{
-					boolResult = _parser.Parse<Request, bool>(rule.Expression, request);
-					_log.LogInformation($"Result of expression {rule.Expression} is {boolResult}");
+                {
+                    var inputType = AppDomain.CurrentDomain.GetAssemblies().
+                        SelectMany(t => t.GetTypes()).First(t => string.Equals(t.Name, rule.InputType, StringComparison.Ordinal));
+                    var method = methods.First(m => m.ReturnType.IsGenericParameter && m.GetGenericArguments().Length == 2);
```
Comment (troyxli) [inline RuleEngine/Application.cs:80]: This finds the generic method V Parse<T, V>(string expression, T model).

File: RuleEngine/Application.cs (hunk 2 above)
Comment (long2know) [reply RuleEngine/Application.cs:65]: It's probably also not necessary - I would imagine 99% of the time, a bool return type would be desired for evaluating a truthy expression.  The "ReturnType" was just a thought on the BusinessRule object.

**Pull request merged**